lightning/src/ln/outbound_payment.rs

```diff
@@ -1288,4 +1312,53 @@ mod tests {
 			assert!(err.contains("Failed to find a route"));
 		} else { panic!("Unexpected error"); }
 	}
+
+	#[test]
+	fn fail_on_retry_error() {
+		let outbound_payments = OutboundPayments::new();
+		let logger = test_utils::TestLogger::new();
+		let genesis_hash = genesis_block(Network::Testnet).header.block_hash();
+		let network_graph = Arc::new(NetworkGraph::new(genesis_hash, &logger));
+		let router = test_utils::TestRouter::new(network_graph);
+		let secp_ctx = Secp256k1::new();
+		let session_priv = [42; 32];
+		let keys_manager = test_utils::TestKeysInterface::new(&[0; 32], Network::Testnet);
+		{
+			let mut random_bytes_override = keys_manager.override_random_bytes.lock().unwrap();
+			*random_bytes_override = Some(session_priv);
+		}
+
+		let payment_params = PaymentParameters::from_node_id(
+			PublicKey::from_secret_key(&secp_ctx, &SecretKey::from_slice(&[42; 32]).unwrap()), 0);
+		let route_params = RouteParameters {
+			payment_params,
+			final_value_msat: 1000,
+			final_cltv_expiry_delta: 0,
+		};
+		router.expect_find_route(route_params.clone(),
+			Err(LightningError { err: String::new(), action: ErrorAction::IgnoreError }));
+		let our_payment_id = PaymentId([0; 32]);
+		let our_payment_hash = PaymentHash([0; 32]);
+		outbound_payments.add_new_pending_payment(our_payment_hash, None, our_payment_id, None,
+			&Route { paths: vec![], payment_params: None }, Some(Retry::Attempts(1)),
+			Some(route_params.payment_params.clone()), &&keys_manager, 0).unwrap();
+		{
+			let mut pending_outbounds = outbound_payments.pending_outbound_payments.lock().unwrap();
+			if let Some(PendingOutboundPayment::Retryable { ref mut total_msat, .. }) = pending_outbounds.get_mut(&our_payment_id) {
+				*total_msat += 1000;
+			}
+		}
```

Comment:
Confused what this block is for. I'm thinking this whole test to make sure that `check_retry_payments` abandons + fails the payment after exhausting retry attempts...is this part to simulate that there are more sats that still need to be sent than are pending..?

Reply:
> is this part to simulate that there are more sats that still need to be sent than are pending..?

Yup, exactly. Added a comment to clarify